### PR TITLE
no-comments option added

### PR DESCRIPTION
**Pull request by berewt**
Add --no-comments option #859

### CODE REVIEW

**Comment by paf31** [comment]
Looks good, but

- where is `removeComments` defined?
- I'd still prefer to pass the options to the code generator. That way, we can change our minds and make no comments the default, or even support different comment styles later on.

Actually, I've been thinking that @michaelficarra is right about making no comments the default, especially since I want to document the Prelude, and the code size is going to blow up.

**Comment by berewt** [comment]
removeComments was already there (Purescript.CoreFn.Ann): it is used to remove the comments from ne annotation when a JSComment is generated.

At this stage, making no-comments the default can be done quite easily (rename no-comments to keep-comments, invert the condition in moduleToJS). Anyway, i will see how to do it at the codeGen phase.

**Comment by berewt** [comment]
Just to clarify: by codegen, you talk about "Language.PureScript.Pretty.JS.prettyPrintJS" or do I miss the point?

**Comment by jdegoes** [comment]
@paf31 I don't think anyone would deploy a Purescript (or *any* Javascript application) without first minifying. That said, I don't imagine myself looking at the JS code for comments.

Now adding comments for Closure compiler, that's another matter.

**Comment by garyb** [comment]
@berewt by codegen he probably means still in `Language.PureScript.Pretty.JS, but by passing the option through to `nonRecToJS` and skipping the generation of `JSComment` there instead, when appropriate.

**Comment by berewt** [comment]
@garyb Ok, I Functor's instances is an overkill ni this context. I might as well remove them and do the change inside `nonRecToJS` then.
Thanks.

**Comment by paf31** [comment]
I need a `Functor` instance for something to do with rewrite rules, so leave them in, they'll come in handy :)

**Comment by berewt** [comment]
Finally found the time to change it, carrying the options through the code generation was a bit more painful than expected.

**Comment by garyb** [comment]
You could have just passed an extra argument from `moduleToJs` through `bindToJs` to `nonRecToJS`. ;)

I have no strong preference of that vs reader though, I'll let @paf31 weigh in.

**Comment by paf31** [comment]
On the one hand, `ReaderT` seems like overkill, for such a short code path. On the other hand, it doesn't seem enough :) This is exactly what we should be doing with the `Options` object IMO, but we should be doing it everywhere. 

I just don't want to be in a situation where we use `ReaderT` in some places and not in others, so for me it's all or none, but I would prefer all.

**Comment by berewt** [comment]
Using Reader was far more convenient for the refactoring as bindToJs is called by many other functions in the file.
I agree that it would be better if it was used everywhere to handle options though.
I will try to change them into parameter passing then.

**Comment by berewt** [comment]
I can also check if I can generalize ReaderT. The weather forecasts are bad for the weekend anyway.

**Comment by garyb** [comment]
Ah good point, I'd forgotten about the calls to `bindToJs` for `let`, etc.

**Comment by berewt** [comment]
Yes, I used Reader in 7 functions I think and I need to pass options in all of them, that's why I move for a Reader instead of parameter passing.

**Comment by berewt** [comment]
Hear is a version with generalized Reader, we can still go a bit Further:

* using `StateT` for the Environment in `PureScript.hs`
* using `ReaderT` in psc / psci for the options

I also tried to do a Reader version of the `JS.hs` changes but it's cumbersome, really.

File: src/Language/PureScript.hs [hunk 1]
```diff
@@ -152,13 +157,14 @@ traverseEither f (Right y) = Right <$> f y
 -- If timestamps have not changed, the externs file can be used to provide the module's types without
 -- having to typecheck the module again.
 --
-make :: (Functor m, Applicative m, Monad m, MonadMake m) => FilePath -> Options Make -> [(Either RebuildPolicy FilePath, Module)] -> [String] -> m Environment
-make outputDir opts ms prefix = do
+make :: (Functor m, Applicative m, Monad m, MonadMake m) => FilePath -> [(Either RebuildPolicy FilePath, Module)] -> [String] -> ReaderT (Options Make) m Environment
```
Comment (paf31) [inline src/Language/PureScript.hs]: I'd prefer to use a `MonadReader` constraint here, and in other places.

File: psci/Main.hs [hunk 2]
```diff
@@ -402,7 +403,7 @@ handleDeclaration :: P.Expr -> PSCI ()
 handleDeclaration val = do
   st <- PSCI $ lift get
   let m = createTemporaryModule True st val
-  e <- psciIO . runMake $ P.make modulesDir options (psciLoadedModules st ++ [(Left P.RebuildAlways, m)]) []
+  e <- psciIO . runMake $ P.make modulesDir (psciLoadedModules st ++ [(Left P.RebuildAlways, m)]) [] `runReaderT` options
```
Comment (paf31) [inline psci/Main.hs]: The `ReaderT` transformer can just be moved into the `PSCI` stack.

**Comment by paf31** [comment]
This already looks _much_ nicer in my opinion. 

Please don't take my comments the wrong way - this is tidying up some stuff which I've wanted to do for a while, and this is just a convenient time since we're touching related code. If you don't want to include it in this PR, that's totally fine, but I think this is a good chance to tidy lots of stuff up, if we're doing the `ReaderT` thing anyway. So that said:

- I would like to use `mtl` classes like `MonadReader` instead of `ReaderT` directly. Many functions already work over an arbitrary monad, so it seems a shame to restrict to `ReaderT` on the outside.
- Eventually, I would like to do the same with a `MonadSupply` to remove the fixed `SupplyT` type constructor.

As I say though, I can merge now. Eventually I'd like to move to the `mtl` style in more places though.

**Comment by berewt** [comment]
Don't worry, I totally understand that your comments wasn't negative but positive suggestions. I totally agree that using typeclasses instead of their instances is better too. 
I can go through the changes and move `ReaderT` to `MonadReader` easily I think.

I will do my best to introduce other generic `mtl` changes too.

Thanks for your feedbacks.

**Comment by berewt** [comment]
I think I would need to introduce FlexibleContexts to do so, is it a problem?

**Comment by paf31** [comment]
Nope, any sensible extensions like that are welcome.

**Comment by berewt** [comment]
You can provide feedback on the style but don't push it yet, I will work on the rest of the Reader code today.

**Comment by paf31** [comment]
Wonderful - thank you very much!

I think you might need to do a bit of merging after the recent optimizer changes, but there shouldn't be much to do.

One last thing - can we please change the flag to `--comments` or `-c`, to keep comments, instead of to remove comments? I think no comments makes a good default and matches other compilers like TypeScript.

I'll work on adding a `MonadSupply` class when this is merged.

**Comment by berewt** [comment]
No problem, done.

**Comment by paf31** [comment]
Looks good to me, thanks! @garyb, any comments?

@berewt Could you please squash the commits?

**Comment by berewt** [comment]
Commits squashed!

**Comment by garyb** [comment]
I do have a comment: Thanks! Looks good to me :+1:

**Comment by berewt** [comment]
I amended the last commit to add the -c alternative.
And many thanks to both of you.

**Comment by paf31** [comment]
This is really great, thanks!

**Pull request merged**